server/auth/auth.go

```diff
@@ -0,0 +1,240 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package auth
+
+import (
+	"encoding/json"
+	"path"
+	"strings"
+	"sync"
+
+	"github.com/tikv/pd/pkg/errs"
+	"github.com/tikv/pd/server/kv"
+	"go.etcd.io/etcd/clientv3"
+)
+
+// RBACManager is used for the rbac storage, cache, management and enforcing logic.
+type RBACManager struct {
+	roleManager
+}
+
+type roleManager struct {
+	kv    kv.Base
+	mu    sync.RWMutex
+	roles map[string]*Role
+}
+
+// newRoleManager creates a new roleManager.
+func newRoleManager(kv kv.Base) *roleManager {
+	return &roleManager{kv: kv, roles: make(map[string]*Role)}
+}
+
+// GetRole returns a role.
+func (m *roleManager) GetRole(name string) (*Role, error) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	role, ok := m.roles[name]
+	if !ok {
+		return nil, errs.ErrRoleNotFound.FastGenByArgs(name)
+	}
+
+	return role, nil
+}
+
+// GetRoles returns all available roles.
+func (m *roleManager) GetRoles() map[string]*Role {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	return m.roles
+}
+
+// CreateRole creates a new role.
+func (m *roleManager) CreateRole(name string) error {
+	_, err := m.GetRole(name)
+	if err == nil {
+		return errs.ErrRoleExists.GenWithStackByArgs(name)
+	}
+
+	role, err := NewRole(name)
+	if err != nil {
+		return err
+	}
+
+	roleJSON, err := json.Marshal(role)
+	if err != nil {
+		return err
+	}
+	rolePath := path.Join(rolePrefix, name)
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// Add role to kv.
+	err = m.kv.Save(rolePath, string(roleJSON))
+	if err != nil {
+		return err
+	}
+
+	// Add role to memory cache.
+	m.roles[name] = role
+
+	return nil
+}
+
+// DeleteRole deletes a role.
+func (m *roleManager) DeleteRole(name string) error {
+	_, err := m.GetRole(name)
+	if err != nil {
+		return err
+	}
+
+	rolePath := path.Join(rolePrefix, name)
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// Delete role from kv.
+	err = m.kv.Remove(rolePath)
+	if err != nil {
+		return err
+	}
+
+	// Delete role from memory cache.
+	delete(m.roles, name)
+
+	return nil
+}
+
+// SetPermissions sets permissions of a role.
+func (m *roleManager) SetPermissions(name string, permissions map[Permission]struct{}) error {
+	role, err := m.GetRole(name)
+	if err != nil {
+		return err
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	role.Permissions = permissions
+
+	// Update role in kv
+	roleJSON, err := json.Marshal(role)
+	if err != nil {
+		return err
+	}
+	rolePath := path.Join(rolePrefix, name)
+
+	err = m.kv.Save(rolePath, string(roleJSON))
+	if err != nil {
+		return err
+	}
+
+	// No need to update role in memory cache again.
+
+	return nil
+}
+
+// AddPermission adds a permission to a role.
+func (m *roleManager) AddPermission(name string, permission Permission) error {
+	role, err := m.GetRole(name)
+	if err != nil {
+		return err
+	}
+
+	if _, ok := role.Permissions[permission]; ok {
```

Comment:
Do permissions of the role need a mutex?

Reply:
Nice catch! Fixed!